### PR TITLE
Composite Index updates to support combinations of ranges

### PR DESCRIPTION
**Pull request by jwomeara**
This expands upon the initial field index filter work to incorporate that logic into the existing composite index functionality.  The composite index code has also been updated to conditionally combine ranges of multiple terms.

### CODE REVIEW

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java [hunk 1]
```diff
@@ -379,17 +412,16 @@ private void compilePatterns() {
...
         public void addCompositeFields(List<NormalizedContentInterface> compositeFields, Multimap<String,NormalizedContentInterface> eventFields,
                         String compositeFieldName, String replacement, String[] grouping, GroupingPolicy groupingPolicy, boolean allowMissing, String[] fields,
-                        int pos, String startSeparator, String endSeparator, StringBuilder originalValue, StringBuilder normalizedValue,
-                        Map<String,String> markings) {
-            String separator = "";
+                        int pos, StringBuilder originalValue, StringBuilder normalizedValue, Map<String,String> markings, boolean isOverloadedField) {
+            String separator = (pos > 0) ? COMPOSITE_DEFAULT_SEPARATOR : "";
             // append any constants that have been specified
             while (pos < fields.length && isConstant(fields[pos])) {
                 String constant = getConstant(fields[pos++]);
```
Comment (tomnelson) [inline warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java]: if pos can't be negative here (array index out of bounds) ....

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java [hunk 2]
```diff
@@ -379,17 +412,16 @@ private void compilePatterns() {
...
-                originalValue.append(constant);
+                if (!isOverloadedField || (isOverloadedField && pos == 0))
```
Comment (tomnelson) [inline warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java]: ... then pos cannot be zero here. or did i miss something?

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 3]
```diff
@@ -1535,13 +1537,10 @@ public IteratorSetting call() throws Exception {
...
                 
                 try {
-                    if (compressMappings) {
-                        addOption(cfg, QueryOptions.COMPOSITE_METADATA,
-                                        QueryOptions.compressOption(metadataHelper.getCompositeMetadata().toString(), QueryOptions.UTF8), true);
-                    } else {
-                        addOption(cfg, QueryOptions.COMPOSITE_METADATA, metadataHelper.getCompositeMetadata().toString(), true);
-                    }
-                } catch (TableNotFoundException | IOException e) {
+                    CompositeMetadata compositeMetadata = metadataHelper.getCompositeMetadata().filter(config.getQueryFieldsDatatypes().keySet());
```
Comment (tomnelson) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: CompositeMetadata compositeMetadata = metadataHelper.getCompositeMetadata().filter(config.getQueryFieldsDatatypes().keySet());

should be:

CompositeMetadata compositeMetadata = metadataHelper.getCompositeMetadata().filter(config.getCompositeToFieldMap().keySet());

File: warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java [hunk 4]
```diff
@@ -149,77 +162,66 @@ public String getCompositeAttributeKey(String composite, String grouping) {
...
                 Attributes attrs = (Attributes) attr;
                 int newAttributeCount = attrs.size();
-                int originalDataListSize = dataList.length;
+                int originalDataListSize = dataList.size();
                 if (newAttributeCount > 0) {
-                    if (dataList.length == 0) {
-                        dataList = new String[newAttributeCount];
-                        int i = 0;
+                    if (dataList.size() == 0) {
                         for (Attribute<?> a : attrs.getAttributes()) {
```
Comment (tomnelson) [inline warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java]: you're not using Attribute<?> a in this for loop

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java (hunk 1 above)
Comment (jwomeara) [reply warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java]: pos can be 0 though, and in that case we don't need the separator yet.

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java (hunk 2 above)
Comment (jwomeara) [reply warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java]: Yup, this needs to be changed.

File: warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java (hunk 4 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java]: Ooops, I think that was a leftover from the original logic before refactor.

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 3 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: Fixed!

File: warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java (hunk 4 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java]: Fixed!

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java (hunk 2 above)
Comment (jwomeara) [reply warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/CompositeIngest.java]: Fixed!

**Pull request merged**